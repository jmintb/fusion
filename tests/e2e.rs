use anyhow::Result;
use assert_cmd::Command;
use rstest::rstest;
use std::path::PathBuf;

#[rstest]
fn test_well_formed_programs(#[files("./test_programs/*.ts")] path: PathBuf) -> Result<()> {
    let mut cmd = Command::cargo_bin("typed_script")?;
    let assert = cmd.arg("run").arg(path.clone()).assert();
    assert.success();
    insta::assert_snapshot!(
        format!(
            "test_well_formed_programs_{}",
            path.file_name().unwrap().to_str().unwrap()
        ),
        String::from_utf8(cmd.ok().unwrap().stdout)?
    );
    Ok(())
}

#[rstest]
fn snapshop_mlir_output(#[files("./test_programs/*.ts")] path: PathBuf) -> Result<()> {
    let mut cmd = Command::cargo_bin("typed_script")?;
    let assert = cmd
        .args(&["build", "--emit-mlir"])
        .arg(path.clone())
        .assert();
    assert.success();
    insta::assert_snapshot!(
        format!(
            "test_mlir_snapshot_{}",
            path.file_name().unwrap().to_str().unwrap()
        ),
        String::from_utf8(cmd.ok().unwrap().stdout)?
    );
    Ok(())
}

#[rstest]
fn snapshop_ast_output(#[files("./test_programs/*.ts")] path: PathBuf) -> Result<()> {
    let mut cmd = Command::cargo_bin("typed_script")?;
    let assert = cmd
        .args(&["build", "--emit-ast"])
        .arg(path.clone())
        .assert();
    assert.success();
    insta::assert_snapshot!(
        format!(
            "test_ast_snapshot_{}",
            path.file_name().unwrap().to_str().unwrap()
        ),
        String::from_utf8(cmd.ok().unwrap().stdout)?
    );
    Ok(())
}
