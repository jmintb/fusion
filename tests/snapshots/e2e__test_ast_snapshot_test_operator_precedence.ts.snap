---
source: tests/e2e.rs
expression: "String::from_utf8(cmd.ok().unwrap().stdout)?"
---
[
    Decl(
        Function {
            keywords: [
                Normal,
            ],
            id: TSIdentifier(
                "main",
            ),
            arguments: [],
            body: Some(
                [
                    Assignment(
                        Assignment {
                            id: TSIdentifier(
                                "a",
                            ),
                            expression: Value(
                                Integer(
                                    2,
                                ),
                                Integer,
                            ),
                            typed_annotation: None,
                        },
                    ),
                    Assignment(
                        Assignment {
                            id: TSIdentifier(
                                "b",
                            ),
                            expression: Value(
                                Integer(
                                    4,
                                ),
                                Integer,
                            ),
                            typed_annotation: None,
                        },
                    ),
                    Assignment(
                        Assignment {
                            id: TSIdentifier(
                                "c",
                            ),
                            expression: Value(
                                Integer(
                                    3,
                                ),
                                Integer,
                            ),
                            typed_annotation: None,
                        },
                    ),
                    Assignment(
                        Assignment {
                            id: TSIdentifier(
                                "resultb",
                            ),
                            expression: Operation(
                                Binary(
                                    Operation(
                                        Binary(
                                            Value(
                                                Variable(
                                                    TSIdentifier(
                                                        "c",
                                                    ),
                                                ),
                                                Unknown,
                                            ),
                                            Addition,
                                            Value(
                                                Variable(
                                                    TSIdentifier(
                                                        "a",
                                                    ),
                                                ),
                                                Unknown,
                                            ),
                                        ),
                                    ),
                                    Multiplication,
                                    Value(
                                        Variable(
                                            TSIdentifier(
                                                "b",
                                            ),
                                        ),
                                        Unknown,
                                    ),
                                ),
                            ),
                            typed_annotation: None,
                        },
                    ),
                    Assignment(
                        Assignment {
                            id: TSIdentifier(
                                "outputb",
                            ),
                            expression: Value(
                                String(
                                    "resultb: %d \\n",
                                ),
                                String,
                            ),
                            typed_annotation: None,
                        },
                    ),
                    Expression(
                        Call(
                            TSIdentifier(
                                "sprintf",
                            ),
                            [
                                Value(
                                    Variable(
                                        TSIdentifier(
                                            "outputb",
                                        ),
                                    ),
                                    Unknown,
                                ),
                                Value(
                                    Variable(
                                        TSIdentifier(
                                            "outputb",
                                        ),
                                    ),
                                    Unknown,
                                ),
                                Value(
                                    Variable(
                                        TSIdentifier(
                                            "resultb",
                                        ),
                                    ),
                                    Unknown,
                                ),
                            ],
                        ),
                    ),
                    Expression(
                        Call(
                            TSIdentifier(
                                "print",
                            ),
                            [
                                Value(
                                    Variable(
                                        TSIdentifier(
                                            "outputb",
                                        ),
                                    ),
                                    Unknown,
                                ),
                                Value(
                                    Integer(
                                        13,
                                    ),
                                    Integer,
                                ),
                            ],
                        ),
                    ),
                    Expression(
                        Return(
                            Return {
                                expression: None,
                            },
                        ),
                    ),
                ],
            ),
            return_type: Unit,
        },
    ),
]
